File: iosNativeApp/iosNativeApp/Presentation/Screen/Main/MainViewModel.swift
import Foundation

@MainActor
final class MainViewModel: ObservableObject {
    @Published var currencies: [CurrencyUIModel] = []
    @Published var showLoading = false
    @Published var showAddSheet = false
    @Published var availableCurrenciesForAdd: [String] = []
    @Published var showError = false

    private var ratesMap: [String: Double] = [:]
    private var lastEditedCurrency: String? = nil
    private var lastEditedValue: Double = 0.0

    private let getCurrencyRatesUseCase: GetCurrencyRatesUseCase

    init(getCurrencyRatesUseCase: GetCurrencyRatesUseCase) {
        self.getCurrencyRatesUseCase = getCurrencyRatesUseCase
        self.currencies = SettingsController.getCurrencies().map { CurrencyUIModel(name: $0, value: "") }
        Task { await loadRates() }
    }

    func onCurrencyAddClicked() {
        let currentNames = Set(currencies.map { $0.name })
        availableCurrenciesForAdd = ratesMap.keys
            .filter { !currentNames.contains($0) }
            .sorted()
        showAddSheet = true
    }

    func onCurrencyValueChanged(_ model: CurrencyUIModel) {
        guard let inputValue = Double(model.value) else {
            currencies = currencies.map { CurrencyUIModel(name: $0.name, value: "") }
            lastEditedCurrency = nil
            lastEditedValue = 0
            return
        }

        lastEditedCurrency = model.name
        lastEditedValue = inputValue

        guard let rateOfEdited = ratesMap[model.name] else { return }
        let amountInUsd = inputValue / rateOfEdited

        currencies = currencies.map { currency in
            if currency.name == model.name {
                return CurrencyUIModel(name: currency.name, value: model.value)
            } else if let targetRate = ratesMap[currency.name] {
                return CurrencyUIModel(name: currency.name, value: formatValue(amountInUsd * targetRate))
            } else {
                return CurrencyUIModel(name: currency.name, value: "")
            }
        }
    }

    func onCurrencyAdded(_ currency: String) {
        guard !currencies.contains(where: { $0.name == currency }) else { return }
        let newCurrency = CurrencyUIModel(name: currency, value: computeValueForCurrency(currency))
        currencies.append(newCurrency)
        SettingsController.saveCurrencies(Set(currencies.map { $0.name }))
    }

    func onCurrencyDeleted(_ currency: String) {
        currencies.removeAll { $0.name == currency }
        SettingsController.saveCurrencies(Set(currencies.map { $0.name }))
    }

    private func loadRates() async {
        showLoading = true
        let result = await getCurrencyRatesUseCase.execute()
        showLoading = false
        switch result {
        case .success(let rates):
            ratesMap = rates
        case .failure:
            showError = true
        }
    }

    private func computeValueForCurrency(_ currency: String) -> String {
        guard let editedCurrency = lastEditedCurrency,
              let rateOfEdited = ratesMap[editedCurrency],
              let targetRate = ratesMap[currency] else { return "" }
        return formatValue((lastEditedValue / rateOfEdited) * targetRate)
    }

    private func formatValue(_ value: Double) -> String {
        if value >= 1.0 {
            let rounded = Int64(value * 100)
            let intPart = rounded / 100
            let fracPart = rounded % 100
            return "\(intPart).\(String(format: "%02d", fracPart))"
        } else {
            let scaled = Int64(value * 100_000_000)
            var str = String(scaled)
            while str.count < 8 { str = "0" + str }
            while str.hasSuffix("0") { str.removeLast() }
            return str.isEmpty ? "0" : "0.\(str)"
        }
    }
}

File: sharedUI/src/commonMain/kotlin/io/github/currencyconverter/presentation/screen/main/MainViewModel.kt
package io.github.currencyconverter.presentation.screen.main

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import io.github.currencyconverter.domain.usecase.GetCurrencyRatesUseCase
import io.github.currencyconverter.presentation.model.CurrencyUIModel
import io.github.currencyconverter.tools.SettingsController
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class MainViewModel(
    private val getCurrencyRatesUseCase: GetCurrencyRatesUseCase
) : ViewModel() {

    private val _state: MutableStateFlow<MainScreenState> = MutableStateFlow(
        MainScreenState(
            currencies = SettingsController.getCurrencies().map { name ->
                CurrencyUIModel(name = name, value = "")
            },
        )
    )
    private val _event = Channel<MainScreenEvent>(Channel.BUFFERED)

    private var ratesMap: Map<String, Double> = emptyMap()
    private var lastEditedCurrency: String? = null
    private var lastEditedValue: Double = 0.0

    val event = _event.receiveAsFlow()
    val state: StateFlow<MainScreenState> = _state.asStateFlow()

    init {
        loadRates()
    }

    fun onIntent(intent: MainScreenIntent) {
        when (intent) {
            is MainScreenIntent.OnCurrencyValueChanged -> onCurrencyValueChanged(intent.model)
            is MainScreenIntent.OnCurrencyAdded -> onCurrencyAdded(intent.currency)
            is MainScreenIntent.OnCurrencyDeleted -> onCurrencyDeleted(intent.currency)
            MainScreenIntent.OnCurrencyAddClicked -> onCurrencyAddClicked()
        }
    }

    private fun loadRates() {
        _state.update { it.copy(showLoading = true) }
        viewModelScope.launch {
            getCurrencyRatesUseCase.invoke()
                .onSuccess { rates ->
                    ratesMap = rates
                    _state.update { it.copy(showLoading = false) }
                }
                .onFailure {
                    _state.update { it.copy(showLoading = false) }
                    viewModelScope.launch { _event.send(MainScreenEvent.ShowError) }
                }
        }
    }

    private fun onCurrencyAddClicked() {
        viewModelScope.launch {
            val currentCurrencies = _state.value.currencies.map { it.name }.toSet()

            _event.send(
                MainScreenEvent.NavigateToAdd(
                    availableCurrencies = ratesMap.keys
                        .filter { it !in currentCurrencies }
                        .sorted()
                )
            )
        }
    }

    private fun onCurrencyValueChanged(model: CurrencyUIModel) {
        val inputValue = model.value.toDoubleOrNull() ?: run {
            _state.update { state ->
                state.copy(
                    currencies = state.currencies.map { it.copy(value = "") }
                )
            }
            lastEditedCurrency = null
            lastEditedValue = 0.0
            return
        }

        lastEditedCurrency = model.name
        lastEditedValue = inputValue

        val rateOfEdited = ratesMap[model.name] ?: return
        val amountInUsd = inputValue / rateOfEdited

        _state.update { state ->
            state.copy(
                currencies = state.currencies.map { currency ->
                    if (currency.name == model.name) {
                        currency.copy(value = model.value)
                    } else {
                        val targetRate = ratesMap[currency.name]
                        if (targetRate != null) {
                            currency.copy(value = formatValue(amountInUsd * targetRate))
                        } else {
                            currency.copy(value = "")
                        }
                    }
                }
            )
        }
    }

    private fun onCurrencyAdded(currency: String) {
        val currentCurrencies = _state.value.currencies
        if (currentCurrencies.any { it.name == currency }) return

        val newCurrency = CurrencyUIModel(
            name = currency,
            value = computeValueForCurrency(currency)
        )
        val updatedList = currentCurrencies + newCurrency

        SettingsController.saveCurrencies(updatedList.map { it.name }.toSet())
        _state.update { it.copy(currencies = updatedList) }
    }

    private fun onCurrencyDeleted(currency: String) {
        val updatedList = _state.value.currencies.filter { it.name != currency }
        SettingsController.saveCurrencies(updatedList.map { it.name }.toSet())
        _state.update { it.copy(currencies = updatedList) }
    }

    private fun computeValueForCurrency(currency: String): String {
        val editedCurrency = lastEditedCurrency ?: return ""
        val rateOfEdited = ratesMap[editedCurrency] ?: return ""
        val targetRate = ratesMap[currency] ?: return ""

        val amountInUsd = lastEditedValue / rateOfEdited
        return formatValue(amountInUsd * targetRate)
    }

    private fun formatValue(value: Double): String {
        return if (value >= 1.0) {
            val rounded = (value * 100).toLong()
            val intPart = rounded / 100
            val fracPart = rounded % 100
            "$intPart.${fracPart.toString().padStart(2, '0')}"
        } else {
            val scaled = (value * 100_000_000).toLong()
            val str = scaled.toString().padStart(8, '0')
            val trimmed = str.trimEnd('0')
            "0.$trimmed".trimEnd('.')
        }
    }
}